Lib/functools.py

```diff
@@ -968,7 +967,11 @@ def __get__(self, obj, cls=None):
                 return _method
 
         dispatch = self.dispatcher.dispatch
+        funcname = getattr(self.func, '__name__', 'singledispatchmethod method')
         def _method(*args, **kwargs):
+            if not args:
+                raise TypeError(f'{funcname} requires at least '
+                                '1 positional argument')
             return dispatch(args[0].__class__).__get__(obj, cls)(*args, **kwargs)
```

Comment:
I haven't done benchmarks, but is there a risk that this could add unnecessary overhead to the happy path? What if we did something like this @serhiy-storchaka?

```suggestion
            try:
                key = args[0].__class__
            except IndexError:
                raise TypeError(f'{funcname} requires at least '
                                '1 positional argument') from None
            return dispatch(key).__get__(obj, cls)(*args, **kwargs)
```

Reply:
It looks reasonable, but could you please benchmark this? Your code is not zero cost either, it adds STORE_FAST+LOAD_FAST instead of LOAD_FAST+TO_BOOL+POP_JUMP_IF_TRUE.

If there is a difference, we perhaps should apply the same optimization in `singledispatch()`.

Reply:
Yeah, I'll benchmark it.

Reply:
For now, I reverted this change because it complicates the code. And the solution for other issues may change it anyway. It is better to defer it until the code will be stabilized.